src/solver/tron.jl

```diff
@@ -0,0 +1,403 @@
+    @printf("%4d  %9.2e  %7.1e  %7.1e\n", iter, fx, πx, get_property(tr, :radius))
+  end
+  while !(optimal || tired || stalled)
+    # Current iteration
+    xc .= x
+    fc = fx
+    Δ = get_property(tr, :radius)
+
+    # Model
+    H = hess_op!(nlp, xc, temp)
+
+    # Cauchy starts here
+    αC, s = cauchy(x, H, gx, Δ, αC, ℓ, u, μ₀=μ₀, μ₁=μ₁, σ=σ)
+
+    s, Hs, cgits, cginfo = projected_newton!(x, H, gx, Δ, cgtol, s, ℓ, u, max_cgiter=max_cgiter)
+    slope = dot(gx,s)
+    qs = 0.5 * dot(s, Hs) + slope
+    fx = f(x)
+
+    try
+      ratio!(tr, nlp, fc, fx, qs, x, s, slope)
+    catch exc
+      status = exc.msg
+      stalled = true
+      continue
+    end
+
+    if num_success_iters == 0
+      tr.radius = min(Δ, norm(s))
+    end
+
+    # Update the trust region
+    update!(tr, norm(s))
+
+    if acceptable(tr)
+      num_success_iters += 1
+      gx = g(x)
+      project_step!(gpx, x, -gx, ℓ, u)
+      πx = norm(gpx)
+    end
+
+    # No post-iteration
+
+    if !acceptable(tr)
+      fx = fc
+      x .= xc
+    end
+
+    iter += 1
+    el_time = time() - start_time
+    tired = iter >= itmax ||  el_time > timemax
+    optimal = πx <= ϵ || fx == -Inf
+
+    verbose && @printf("%4d  %9.2e  %7.1e  %7.1e  %7.1e  %s\n", iter, fx, πx, Δ, Ared / Pred, cginfo)
+  end
+
+  status = if tired
+    iter >= itmax ? "maximum number of iterations" : "maximum elapsed time"
+  elseif πx <= ϵ
+    "first-order stationary"
+  elseif fx == -Inf
+    "objective function is unbounded from below"
+  elseif status != ""
+    status
+  elseif stalled
+    "stalled"
+  else
+    "unknown"
+  end
+
+  return x, fx, πx, iter, optimal, tired, status, el_time
+end
+
+"""
+    nbrk, brkmin, brkmax = breakpoints(x, d, ℓ, u)
+
+Find the smallest and largest values of `α` such that `x + αd` lies on the
+boundary. `x` is assumed to be feasible. `nbrk` is the number of breakpoints
+from `x` in the direction `d`.
+"""
+function breakpoints(x::Vector, d::Vector, ℓ::Vector, u::Vector)
+  n = length(x)
+  pos = find( (d .> 0) .& (x .< u) )
+  neg = find( (d .< 0) .& (x .> ℓ) )
+
+  nbrk = length(pos) + length(neg)
+  nbrk == 0 && return 0, 0.0, 0.0
+
+  brkmin = Inf
+  brkmax = 0.0
+  if length(pos) > 0
+    @views steps = (u[pos] - x[pos]) ./ d[pos]
+    brkmin = min.(brkmin, minimum(steps))
+    brkmax = max.(brkmax, maximum(steps))
+  end
+  if length(neg) > 0
+    @views steps = (ℓ[neg] - x[neg]) ./ d[neg]
+    brkmin = min.(brkmin, minimum(steps))
+    brkmax = max.(brkmax, maximum(steps))
+  end
+  return nbrk, brkmin, brkmax
+end
+
+"""`slope, qs = compute_Hs_slope_qs!(Hs, H, s, g)`
+
+Computes
+
+    Hs = H * s
+    slope = dot(g,s)
+    qs = ¹/₂sᵀHs + slope
+"""
+function compute_Hs_slope_qs!(Hs::Vector, H::Union{AbstractMatrix,AbstractLinearOperator},
+                              s::Vector, g::Vector)
+  Hs .= H * s
+  slope = dot(g,s)
+  qs = 0.5 * dot(s, Hs) + slope
+  return slope, qs
+end
+
+"""`active(x, ℓ, u; rtol = 1e-8, atol = 1e-8)`
```

Comment:
We tend to use 1.0e-8 because we're thinking Float64, but your functions are more general. Perhaps 1.0e-8 should be `sqrt(eps(eltype(x)))`?!